api/spec/src/entitlements/entitlements.tsp

```diff
@@ -0,0 +1,319 @@
+   */
+  @example("{ \"integrations\": [\"github\"] }")
+  @encode("json")
+  config: string;
+}
+
+/**
+ * Create inputs for boolean entitlement
+ */
+@friendlyName("EntitlementBooleanCreateInputs")
+model EntitlementBooleanCreateInputs {
+  ...EntitlementCreateSharedFields;
+  type: EntitlementType.boolean;
+}
+
+/**
+ * Create inputs for entitlement
+ */
+@discriminator("type")
+@friendlyName("EntitlementCreateInputs")
+union EntitlementCreateInputs {
+  metered: EntitlementMeteredCreateInputs,
+  static: EntitlementStaticCreateInputs,
+  boolean: EntitlementBooleanCreateInputs,
+}
+
+/**
+ * Shared fields of the entitlement templates.
+ */
+@friendlyName("EntitlementBaseTemplate")
+model EntitlementSharedFields {
+  ...ResourceTimestamps;
+  ...OmitProperties<EntitlementCreateSharedFields, "featureKey" | "featureId">;
+
+  /**
+   * The type of the entitlement.
+   */
+  @summary("Type")
+  type: EntitlementType;
+
+  /**
+   * The identifier key unique to the subject
+   */
+  @example("customer-1")
+  subjectkey: Key;
+
+  /**
+   * The feature the subject is entitled to use.
+   * Either featureKey or featureId is required.
+   */
+  @example("example-feature-key")
+  featureKey: Key;
+
+  /**
+   * The feature the subject is entitled to use.
+   * Either featureKey or featureId is required.
+   */
+  @example("01ARZ3NDEKTSV4RRFFQ69G5FAV")
+  featureId: ULID;
+
+  currentUsagePeriod: Period;
+}
+
+/**
+ * Calculated fields for metered entitlements.
+ */
+@friendlyName("EntitlementMeteredCalculatedFields")
+model EntitlementMeteredCalculatedFields {
+  /**
+   * The time the last reset happened.
+   */
+  @example(DateTime.fromISO("2023-01-01T01:01:01.001Z"))
+  @visibility("read")
+  lastReset: DateTime;
+
+  @visibility("read")
+  currentUsagePeriod: Period;
+
+  /**
+   * The time from which usage is measured. If not specified on creation, defaults to entitlement creation time.
+   */
+  @visibility("read")
+  measureUsageFrom: DateTime;
+}
+
+/**
+ * Metered entitlements are useful for many different use cases, from setting up usage based access to implementing complex credit systems.
+ * Access is determined based on feature usage using a balance calculation (the "usage allowance" provided by the issued grants is "burnt down" by the usage).
+ */
+@friendlyName("EntitlementMetered")
+model EntitlementMetered {
+  ...OmitProperties<
+    EntitlementMeteredCreateInputs,
+    "type" | "measureUsageFrom" | "metadata" | "usagePeriod" | "featureKey" | "featureId" | "currentUsagePeriod"
+  >;
+  ...OmitProperties<EntitlementMeteredCalculatedFields, "currentUsagePeriod">;
+  ...OmitProperties<EntitlementSharedFields, "type">;
+  type: EntitlementType.metered;
+}
+
+/**
+ * Entitlement template of a boolean entitlement.
+ */
+@friendlyName("EntitlementBoolean")
+model EntitlementBoolean extends EntitlementSharedFields {
+  ...OmitProperties<EntitlementBooleanCreateInputs, "type" | "metadata" | "usagePeriod" | "featureKey" | "featureId">;
```

Comment:
There is also PickProperties

Reply:
The docs say there's something analogous to `keyof` which is what i'd like ideally. (I have to exclude an entire subtype shared by multiple composits)

Reply:
The docs say there's something analogous to `keyof` which is what i'd like ideally. (I have to exclude an entire subtype shared by multiple composits)